File: network/tools/ruby/echo_server.rb
## 2022-10-01 K.OHWADA
# echo_server.rb
# Messages to be sent and received 
# are separated by newline code

# https://ruby-doc.org/stdlib-2.1.0/libdoc/socket/rdoc/Socket.html


require 'socket'

PORT = 2000

server = TCPServer.new PORT

printf("listen: %d \n", PORT)

RECV_LENGTH = 100
SEND_FLAGS = 0
SLEEP = 5 # 5 sec
LIMIT_NO_RECV = 10

cnt_no_recv = 0

loop do
    client = server.accept
    addr = client.getpeername
    printf("client: ");
    p Socket.unpack_sockaddr_in(addr)
  while buffer = client.recv(RECV_LENGTH)
        if (buffer.empty?)
# assume that the communication has ended.
# close the socket
# if nothing is received for a certain period of time
            sleep(SLEEP)
            cnt_no_recv += 1
            # printf("  cnt_no_recv: %d \n",   cnt_no_recv)
            if (cnt_no_recv > LIMIT_NO_RECV)
                printf("no recv data \n")
                break
            end # if
        else
            cnt_no_recv += 0
            printf("r: %s \n", buffer.strip)
# echo back
            client.send(buffer, SEND_FLAGS)
        end # if
  end  # while
    printf("close socket \n")
    client.close
    printf("wait new client \n")
end  # loop

File: network/tools/perl/echo_server.pl
#!/usr/local/bin/perl -w
## 2022-10-01 K.OHWADA

# echo_server.pl
# Messages to be sent and received 
# are separated by newline code
# http://x68000.q-e-d.net/~68user/net/echo-2.html

# $Id: echo-server.pl,v 1.2 2002/02/17 11:07:27 68user Exp $


# remove the trailing newline code
sub rtrim {
	my $val = shift;
	$val =~ s/\s*$//;
	return $val;
}

use Socket;

my $port = 5000;

# genarate socket
socket(CLIENT_WAITING, PF_INET, SOCK_STREAM, 0)
     or die "cannot generate socket $!";

# Socket option setting
setsockopt(CLIENT_WAITING, SOL_SOCKET, SO_REUSEADDR, 1)
     or die "setsockopt failed $!";

# assign an address (name) to socket
bind(CLIENT_WAITING, pack_sockaddr_in($port, INADDR_ANY))
     or die "bind failed $!";

# listen port
listen(CLIENT_WAITING, SOMAXCONN)
     or die "listen: $!";

print "listen $port \n";

# by using while(1), prepare for the next connection even if one connection ends
while (1){
    $paddr = accept(CLIENT, CLIENT_WAITING);

# get host name, IP address and client port number
    ($client_port, $client_iaddr) = unpack_sockaddr_in($paddr);

    $client_hostname = gethostbyaddr($client_iaddr, AF_INET);

    $client_ip = inet_ntoa($client_iaddr);

    print "connect: $client_hostname ($client_ip) port $client_port\n";


# don't buffer for clients
# https://perlzemi.com/blog/20080812121854.html
    select(CLIENT); $|=1; select(STDOUT);

    while (my $rcv_data = <CLIENT>){
        my $data = rtrim($rcv_data);
        print "r: $data \n";
#send back a message to the client
        print CLIENT $rcv_data;
    } # while
    close(CLIENT);

    print "Connection lost. Still watching port $port. \n";
}
 # while ---

